test/cql-pytest/test_compaction_tombstone_gc.py

```diff
@@ -0,0 +1,141 @@
+            cql.execute(f"insert into {table} (k, v) values (1, 1) using timestamp 1 and ttl 1")
+            cql.execute(f"insert into {table} (k, v) values (1, 2) using timestamp 2")
+            nodetool.flush(cql, table)
+            # Delete row 2, this deletion is expected to be kept
+            # when we insert backdated live data into the memtable
+            cql.execute(f"delete from {table} using timestamp 3 where k=1 and v=2")
+            # Flush all tables explicitly now, since this is skipped in the test on purpose by next major compaction.
+            nodetool.flush_all(cql)
+
+            # let gc_grace_seconds and ttl pass
+            time.sleep(2)
+            # Re-insert backdated data into the memtable.  It should inhibit tombstone_gc
+            cql.execute(f"insert into {table} (k, v) values (1, 3) using timestamp 2")
+            # do not flush before major compaction
+            nodetool.compact(cql, table, flush_memtables=False)
+
+            res = cql.execute(f"select * from mutation_fragments({table})")
+            sstables = set()
+            keys = set()
+            rows = set()
+            for r in res:
+                if "sstable" in r.mutation_source:
+                    if r.mutation_fragment_kind == "partition start":
+                        sstables.add(r.mutation_source)
+                        keys.add(r.k)
+                    elif r.mutation_fragment_kind == "clustering row":
+                        rows.add(r.v)
+                        if r.v == 2:
+                            assert "tombstone" in r.metadata
+
+            assert len(sstables) == 1, f"Expected single sstable but saw {len(sstables)}: res={list(res)}"
+            assert keys == {1}, f"Expected keys=={1} but got {keys}: res={list(res)}"
+            assert rows == {2}, f"Expected rows=={2} but got {rows}: res={list(res)}"
+
+def test_tombstone_gc_with_delete_in_memtable(scylla_only, cql, test_keyspace):
+    """
+    Reproduce https://github.com/scylladb/scylladb/issues/20423
+    """
+    schema = "k int, v int, primary key (k, v)"
+    with new_test_table(cql, test_keyspace, schema, extra="with gc_grace_seconds = 0") as table:
+        with nodetool.no_autocompaction_context(cql, table):
+            # Insert initial data into the base table
+            cql.execute(f"insert into {table} (k, v) values (1, 1) using timestamp 1 and ttl 1")
+            cql.execute(f"insert into {table} (k, v) values (1, 2) using timestamp 2")
+            cql.execute(f"insert into {table} (k, v) values (1, 3) using timestamp 3")
+            nodetool.flush(cql, table)
+            cql.execute(f"delete from {table} using timestamp 4 where k=1 and v=2")
+            # Flush all tables explicitly now, since this is skipped in the test on purpose by next major compaction.
+            nodetool.flush_all(cql)
+
+            # let gc_grace_seconds and ttl pass
+            time.sleep(2)
```

Comment:
A copy-pasto? There is no TTL in this test, and I gc_grace_seconds is 0 seconds, why do you need to sleep for that to pass? Let's try to avoid sleep() as much as we can.

Reply:
No, it's not about the ttl, it's the gc_grace_seconds and commitlog min_gc.
I'll see if sleep(1) is sufficient here.

(the reason we need sleep(2) for ttl=1 is due to quantization into single seconds)

Reply:
I'll repeat my question, is this a copy-pasto from a test that had ttl=1? Because as far as I saw (maybe I just misread?) this specific place did not have ttl.

Reply:
No, there is an insert using ttl in this test case too.